skbio/sequence/_interval.py

```diff
@@ -0,0 +1,161 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2013--, scikit-bio development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+from __future__ import absolute_import, division, print_function
+from future.utils import with_metaclass
+
+import abc
+
+from ._feature import Feature
+from .intersection import Interval, IntervalTree
+import itertools
+from collections import defaultdict
+
+
+class IntervalMetadata():
+    def __init__(self, features=None):
+        # maps features attributes to intervals
+        if features is None:
+            self.feature_metadata = {}
+            self.interval_metadata = IntervalTree()
+        else:
+            self.interval_metadata = IntervalTree()
+            for k, invs in features.items():
+                for inv in invs:
+                    start, end = _polish_interval(inv)
+                    self.interval_metadata.add(start, end, k)
+            self.feature_metadata = features
+
+    def add(self, x, *intervals):
+        """ Adds a feature to the metadata object.
+
+        Parameters
+        ----------
+        x : skbio.sequence.feature
+            The feature object being added.
+        intervals : iteerable of intervals
+            A list of intervals associated with the feature
+
+        Note
+        ----
+        The intervals associated with a feature is assumed to be under
+        the `location` keyword by default.
+        """
+        if not isinstance(x, Feature):
+            raise ValueError('x is not an instance of `Feature`')
+
+        # TODO: The below will require careful consideration
+        # since this will be using a little more memory, since the whole
+        # feature object will be stored.
+        for interval in intervals:
+            loc = _polish_interval(interval)
+            if loc is not None:
+                start, end = loc
+                self.interval_metadata.add(start, end, x)
+
+        # TODO: The below will require careful consideration
+        # since this will be using a little more memory, since the whole
+        # feature object will be stored.
+        if x not in self.feature_metadata.keys():
+            self.feature_metadata[x] = []
+
+        self.feature_metadata[x] = list(map(_polish_interval, intervals))
+
+
+    def _query_interval(self, interval):
```

Comment:
This gives duplicate if the given interval overlapping with multiple intervals of the same future. should only give one.
>>> di._query_interval((1, 5))
[gene:sagA, gene:sagB, gene:sagA]